decidim/decidim
# API changes to support profile privacy options

#### :tophat: What? Why?
In order to support profile privacy in the external module we are building right now, we will add any proposed API changes to this PR.

This is the second iteration after the following findings:
1. It would be beneficial also for the core to consider the user profile publicity based on if the user has accepted the terms of service which might be a requirement for displaying their profile on the website when they first sign in to the platform.
2. It seems a better approach to consider the profile publicity only in places where we would actually display the user data publicly on the website. This approach seems to require changes in less modules (including 3rd party modules) making it more convenient overall.

~~This is still work in progress but I will open this already now so that we can continue on the same PR.~~

This is now ready for review.

#### :pushpin: Related Issues
- Supersedes #10939
- Related to
  * https://meta.decidim.org/processes/roadmap/f/122/proposals/16247
  * https://meta.decidim.org/processes/roadmap/f/122/proposals/17297
  * https://github.com/mainio/decidim-module-privacy

#### Testing

- Register to the platform as a new user
- See that straight after the registration, you cannot access any of the user's public profile pages (should give a 404)
- See that the user is not visible in the API
- See that managed users do not appear in the API
- Confirm the newly created user profile using the link in the email
- See that you can now access the profile page and see the user details through the API
- Manually change the user in the database to have not accepted the terms of service
- See that the public profile pages are again hidden and the user is not visible through the API

To test what happens in case all user profiles are not published:

- Make a local change to the code: change the [`profile_published?`](https://github.com/decidim/decidim/blob/aa3eb43f655f7fc36830cf193dbce83b8ee1dc92/decidim-core/app/models/decidim/user.rb#L92) method to return `false` always
- See that you can no longer see the details of these users on the website
- See that the "public profile" link is now gone under the user's personal links
- See that the user details are hidden from the API
  * Note that you will still get a list of users through the API using this method because the API query still returns the whole list of confirmed users but the details of these users should be hidden even in this case
